Library/Homebrew/api/formula_hash.rb

```diff
@@ -0,0 +1,224 @@
+
+    class HeadURLHash < T::Struct
+      include CompactSerializable
+
+      const :url, String
+      const :branch, T.nilable(String)
+      const :using, T.nilable(String)
+    end
+
+    # TODO: simplify this
+    class RequirementHash < T::Struct
+      include CompactSerializable
+
+      const :name, String
+      const :cask, T.nilable(String)
+      const :download, T.nilable(String)
+      const :version, T.nilable(String)
+      const :contexts, T::Array[T.untyped]
+      const :specs, T::Array[T.untyped]
+    end
+
+    class ServiceHash < T::Struct
+      include CompactSerializable
+
+      const :name, String
+      const :run, T.nilable(String)
+      const :run_type, T.nilable(String)
+      const :interval, T.nilable(Integer)
+      const :cron, T.nilable(String)
+      const :keep_alive, T.nilable(T::Boolean)
+      const :launch_only_once, T.nilable(T::Boolean)
+      const :require_root, T.nilable(T::Boolean)
+      const :environment_variables, T.nilable(T::Hash[String, String])
+      const :working_dir, T.nilable(String)
+      const :root_dir, T.nilable(String)
+      const :input_path, T.nilable(String)
+      const :log_path, T.nilable(String)
+      const :error_log_path, T.nilable(String)
+      const :restart_delay, T.nilable(Integer)
+      const :process_type, T.nilable(String)
+      const :macos_legacy_timers, T.nilable(T::Boolean)
+      const :sockets, T.nilable(T::Array[String])
+    end
+
+    class StableURLHash < T::Struct
+      include CompactSerializable
+
+      const :url, String
+      const :tag, T.nilable(String)
+      const :revision, T.nilable(String)
+      const :using, T.nilable(String)
+      const :checksum, T.nilable(String)
+    end
+
+    class UsesFromMacOSHash < T::Struct
+      include CompactSerializable
+
+      const :context, T.nilable(Symbol)
+      const :since, T.nilable(String)
+    end
+
+    class FormulaHash < T::Struct
+      include CompactSerializable
+      extend StringifySymbols
+
+      DEPENDENCY_CONTEXTS = T.let([:build, :test, :recommended, :optional].freeze, T::Array[Symbol])
+
+      PROPERTIES = T.let({
+        aliases:                         T::Array[String],
+        bottle:                          BottleHash,
+        caveats:                         T::Array[String],
+        conflicts_with:                  T::Array[String],
+        conflicts_with_reasons:          T::Array[String],
+        deprecation_date:                String,
+        deprecation_reason:              String,
+        deprecation_replacement_cask:    String,
+        deprecation_replacement_formula: String,
+        desc:                            String,
+        disable_date:                    String,
+        disable_reason:                  String,
+        disable_replacement_cask:        String,
+        disable_replacement_formula:     String,
+        head_dependencies:               T::Array[DependencyHash],
+        head_url:                        HeadURLHash,
```

Comment:
@Rylan12 I think we could definitely inline any of these types that aren't in e.g. arrays